### ISSUE TITLE
ARG command and --build-arg not supported

### ISSUE THREAD

**Issue by IlyaSemenov** [grammarly/rocker#104]
Original `docker build` supports providing arguments with `ARG` command and `--build-arg`. Moreover, it allows passing certain arguments (such as `http_proxy`) even without explicitly declaring them in the Dockerfile. (This is so convenient that I even have a bash proxy named `docker` which invokes the real `docker` and passes the host as the http proxy all the time.)

Unfortunately, this is not supported by `rocker build`. Please check the test case below:

```ShellSession
$ cat > Dockerfile << EOF
> FROM busybox
> RUN export | grep proxy
> EOF
$ docker build --build-arg http_proxy=http://host:3128 .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox
 ---> 47bcc53f74dc
Step 2 : RUN export | grep proxy
 ---> Running in 7f2222891324
export http_proxy='http://host:3128'
 ---> 2930972be320
Removing intermediate container 7f2222891324
Successfully built 2930972be320
$ cp Dockerfile Rockerfile
$ rocker build --build-arg http_proxy=http://host:3128
Incorrect Usage.
$ rocker build --var http_proxy=http://host:3128
INFO[0000] FROM busybox
INFO[0000] | Image sha256:47bcc                          size=1.113 MB
INFO[0000] RUN export | grep proxy
INFO[0000] | Not cached
INFO[0000] | Created container bee94d227778 (image sha256:47bcc)
export no_proxy='*.local, 169.254/16'
INFO[0000] Commit changes
INFO[0000] | Result image is sha256:ea55e                size=1.113 MB (+0 B)
INFO[0000] | Removing container bee94d227778
INFO[0000] Cleaning up
INFO[0000] Successfully built sha256:ea55e | final size 1.113 MB (+0 B from the base image)
```

Also, the `ARG` command in the Rockerfile crashes rocker with **FATA[0000] Unknown command: arg**

Yes, I know that rocker has `--var` but it's not really the same:
* It either persists in the image (with `ENV`) or must be used explicitly for each `RUN`.
* It emits `<no value>` (the actual text string) for when the corresponding var has not been provided.
* It requires explicit use while the proxy args in Dockerfile work without declaring them.

I am opening this issue for two reasons:

1) The documentation says: *Yes, you can take any Dockerfile, rename it to `Rockerfile` and use `rocker build` instead of `docker build`* **(No, you can't.)**

2) This limitation kind of contradicts the spirit of Rocker pointed towards ease and reusability. `MOUNT` is great, but there are other means for caching for different scenarios, and using a local http proxy is one of them.

**Comment by diunko** [comment]
Hey @IlyaSemenov! Thanks for reporting this.

We don't currently plan to use ARG commands in {D,R}ockerfile[s], but it looks like a compatibility issue to other people. PR is welcome! :)

**Comment by IlyaSemenov** [comment]
I see. Is there a recommended way to use a local proxy without compromising on Rockerfile portability? It saves tremendous amount of time when developing Ubuntu-based images which run `apt-get` all the time.

**Comment by missedone** [comment]
hi @IlyaSemenov 
does it work if you set the `HTTP_PROXY` environment variable?
see explanation here: http://stackoverflow.com/a/14663620/4867232

I don't have a working http proxy server at hand, but just tried with a dummy one, and was expecting the failure:
```
# this is a dummy proxy
$ export HTTP_PROXY=http://127.0.0.1:1234

$ rocker build
INFO[0000] FROM golang:1.6
INFO[0000] | Image sha256:78e38                          size=753.6 MB
INFO[0000] ADD https://github.com/lalyos/docker-upx/releases/download/v3.91/upx /bin/upx
INFO[0000] Downloading `https://github.com/lalyos/docker-upx/releases/download/v3.91/upx` into `/Users/nick/.rocker_cache/url_fetcher_cache/08/08c257e2ba0943604b7901dfdb344c6a587b30cbb4a8f2ed908060de722dcd2b`
FATA[0000] Get https://github.com/lalyos/docker-upx/releases/download/v3.91/upx: http: error connecting to proxy http://127.0.0.1:1234: dial tcp 127.0.0.1:1234: getsockopt: connection refused
```

**Comment by IlyaSemenov** [comment]
@missedone Thanks, but I don't need Rocker to use the proxy; I need the container to use it during the build process (when it runs `apt-get install`, for instance). Neither `HTTP_PROXY` nor `http_proxy` gets its way into the build container environment, I just tested it.

**Comment by tahajahangir** [comment]
I encountered this issue when I renamed a Dockerfile to Rockerfile and *it doesn't work*! I think this should be fixed as rocker claims to be compatible with Dockerfile.

In the other hand `ARG` is the *only* way to *define* build time variables (possibly with defaults). This feature is now missing in Rockerfile.

**Comment by missedone** [comment]
let me try to work out a PR, will update you later

**Comment by ybogdanov** [comment]
In Rocker 1.3.0

**Comment by skolsuper** [comment]
I am having trouble with the following, using Rocker `1.3.0`:

```Dockerfile
# development / uat / production
ARG deployenv=development

ENV STATICFILES_DIR=/www/data/static \
    DEPLOY_ENVIRONMENT=$deployenv
```

Using `docker build --build-arg deployenv=uat .`, the next RUN command has `$DEPLOY_ENVIRONMENT=$deployenv=uat`, whereas with Rocker the next RUN command has `$deployenv=development` and `$DEPLOY_ENVIRONMENT` not set (well, set to empty string).

**Comment by skolsuper** [comment]
FWIW, I worked around this by changing to:

```dockerfile
# development / uat / production
ARG DEPLOY_ENVIRONMENT=development

RUN mkdir -p /www/data/static
ENV STATICFILES_DIR=/www/data/static
```

Maybe my first version is overly verbose anyway

**Comment by mrfelton** [comment]
This still doesn't seem to work in the sae way that it works in Docker. With docker, you can do something like this:

```
docker build --build-arg NPM_TOKEN --build-arg NODE_ENV .
```

where `NPM_TOKEN` and `NODE_ENV` will be picked up from the currently running environment and passed through docker at build time.

With Rocker, it doesn't appear to be picking up those values from the current environment. Instead, you have to explicitly pass them through as part of the command, meaning that you need to run the command like:

```
rocker build --build-arg NPM_TOKEN=e3115c0c-ddef-4702-8132-eea6438e8f13 --build-arg NODE_ENV=production .
```

This is awkward and not possible or desirable in a lot of situations.

**Comment by ybogdanov** [comment]
@skolsuper @mrfelton thanks for reporting. Fixed in #145

**Comment by ybogdanov** [comment]
By the way, I spent 2 hours in total reading Docker's source code and couldn't find how they take ENV variables in case `--build-arg` cli value is empty.
